### ISSUE TITLE
What data is inherited by RAP for HypreParMatrices

### ISSUE THREAD

**Issue by kvoronin** [mfem/mfem#569]
Hello,

I am using RAP for HypreParMatrices, and at some point I delete the R,A and P matrices.
My code is like the following:
```
block_ij = RAP(TrueP_i, Fine_blk_ij, TrueP_i);
block_ij->CopyRowStarts();
block_ij->CopyColStarts();
```
Is it possible that I need to copy some other data, to have block_ij existing completely independent from R, A and P?

Thanks, 
Kirill

**Comment by v-dobrev** [comment]
I don't think there's anything else you need to copy - `block_ij` should be fine even if you delete all the `HypreParMatrix`es that it was built from.

**Comment by kvoronin** [comment]
Well, the leak and weird results went away right after I replaced rap with two parmults (and deleted the temporary matrix created after the first of them).

**Comment by v-dobrev** [comment]
I don't follow, what are the "leak and weird results"? You did not mention them in the initial post of the issue.

**Comment by kvoronin** [comment]
I'm sorry @v-dobrev, the setup is difficult to explain in detail. 
I started with a memory leak reported for the memory allocated by RAP. I found that the corresponding pointer was overwritten later by HypreParMatrix::Transpose (). But when I added delete for the result of RAP (which was used to create other HypreParMatrices) before overwriting call to Transpose() I started getting wrong matrix dimensions for one of the ParMults which followed Transpose.
The matrix dimension (width and height as a part of Operator) were either 0, or larger than it should be. With that, the result does contain meaningful data as its diagonal block (SparseMatrix).
I don't have time now to study the issue, but the workaround with replacing rap by parmults eliminates the issues - no memory leak, correct results. 
The same thing worked for a few other places where RAP for HypreParMatrices was used.
I checked that everywhere where a new HypreParMatrix was created, I called CopyColStarts and CopyRowStarts.

I merely hope that if someone gets a similar problem in the future and finds this thread on github, one can get away using a hint from here, which is "try to go with two parmults instead of rap".
